src/cpu/instr.rs

```diff
@@ -40,30 +94,230 @@ struct Operation {
                 cpu.state.pc = cpu.state.pc+1;
             },
             repr: |_enc| String::from("nop"),
-        }); 
+        });
+        m.insert(Opcode::MOV as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize];
+            },
+            repr: |enc| format!("mov r{}, r{}", enc.rd, enc.rs1),
+        });
         m.insert(Opcode::LDD as u8, Operation {
             execute: |enc, cpu| {
                 cpu.state.reg[enc.rd as usize] = cpu.read(enc.imm, true);
                 cpu.state.pc = cpu.state.pc+1;
             },
             repr: |enc| format!("ldd r{0}, {1}", enc.rd, enc.imm),
         });
-        m.insert(Opcode::ADD as u8, Operation {
-            execute: |enc, cpu| {
-                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] + cpu.state.reg[enc.rs2 as usize];
+        m.insert(Opcode::ADD as u8, Operation { //changed implementation, because then gen_flag function would loose bits and wouldn't work.
+            execute: |enc, cpu|{
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 + cpu.state.reg[enc.rs2 as usize] as u32; //this gives us acces to 17th bit
+                cpu.state.reg[enc.rd as usize] = _out as u16; // but the same value as before goes back to register
+                cpu.state.flags = gen_flag(false, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs2 as usize] as u32, _out); 
+                //happy genflag function here^
                 cpu.state.pc = cpu.state.pc+1;
             },
             repr: |enc| format!("add r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
         });
+        m.insert(Opcode::ADI as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 + enc.imm as u32;
+                cpu.state.reg[enc.rd as usize] = _out as u16;
+                cpu.state.flags = gen_flag(false, cpu.state.reg[enc.rs1 as usize] as u32, enc.imm as u32, _out);
+
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("adi r{0}, r{1}, {2}", enc.rd, enc.rs1, enc.imm),
+        });
+        m.insert(Opcode::SUB as u8, Operation {
+            execute: |enc, cpu| { 
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 - cpu.state.reg[enc.rs2 as usize] as u32;
+                cpu.state.reg[enc.rd as usize] = _out as u16;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs1 as usize] as u32, _out);
+
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("sub r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::AND as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 & cpu.state.reg[enc.rs2 as usize] as u32;
```

Comment:
That's debatable, but I think that we shouldn't produce _out as u32 in cases where it is not needed (everywhere apart from ADD, ADI, SUB, ADC, SUC, SHL, SHR, CMP, CMI). In normal cases _out can u16, and the only cast is needed in gen_flag call.